runtime/lua/vim/lsp/handlers.lua

```diff
@@ -5,10 +5,21 @@ local util = require('vim.lsp.util')
 local api = vim.api
 local completion = require('vim.lsp.completion')
 
---- @type table<string,lsp.Handler>
+--- @type table<string, lsp.Handler>
 local M = {}
 
--- FIXME: DOC: Expose in vimdocs
+--- @deprecated
+--- Client to server response handlers.
+--- @type table<vim.lsp.protocol.Method.ClientToServer, lsp.Handler>
+local RCS = {}
```

Comment:
I would appreciate a comment explaining why these are indexed by string literals instead of `vim.lsp.protocol.Methods`. Without the context from this PR I would certainly try to change that.

Reply:
They are indexed with `vim.lsp.protocol.Method.ClientToServer` so it is type checked and LuaLS has a stronger understanding of literals. If you use a symbol, then it doesn't understand it as much, or at least that's what I've found.

I also never understood the use of the enum here, it just seems to add a level of indirection. In a statically compiled language it would make some sense since that's usually how you achieve type checking, however in a dynamic language I'm not sure what benefit it provides. It's not like it prevents typos or anything. Was there a reason for adding them I haven't understood?

Reply:
> however in a dynamic language I'm not sure what benefit it provides. It's not like it prevents typos or anything. Was there a reason for adding them I haven't understood?

The completion you get with `protocol.Methods.|` is a nice aspect, and using the enum for table indexing works too - although we should probably account for off-spec extensions in some areas. E.g. not sure if we want to cause warnings for `lsp.handlers["custom"] = ...`

Reply:
You do get completion with this too in LuaLS since the index type is `vim.lsp.protocol.Method.ClientToServer`. You won't get that completion in a `:` command line, but is anyone really using that for this?

> E.g. not sure if we want to cause warnings for lsp.handlers["custom"] = ...

I think `@diagnostic disable` is a pretty good use case for that. Otherwise `lsp.handlers` can be aliased.

Reply:
> You get completion with this too in LuaLS since the index type is vim.lsp.protocol.Method.ClientToServer.

Maybe it's my luals settings, but it appears to me that with enum's it is more reliable.
At least after typing `protocol.Methods.|` I get only the enum options.
With `RSC["|` I see lots of other things too.

I don't mind either way in this case - just pointed it out since you mentioned that you don't see the benefit.

Reply:
I do get something similar:

<img width="556" alt="image" src="https://github.com/user-attachments/assets/7506e41c-0aa1-4b61-8e74-05600a3d23e0">

but the completions are at least there.

I wonder if there's a reason for LuaLS providing the other completions, or maybe it's just a bug.

Reply:
What I meant with my original comment is that adding an explanation to the code explaining why this bit differs from the convention of using the enum as it’s done everywhere else.

Reply:
Decided to revert this since the enums have hover information and the type checking with either isn't that reliable.